provisioner/node_pools.go

```diff
@@ -123,13 +188,124 @@ func (p *AWSNodePoolProvisioner) Provision(ctx context.Context, nodePools []*api
 	return nil
 }
 
-func supportsT2Unlimited(instanceTypes []string) bool {
-	for _, instanceType := range instanceTypes {
-		if !strings.HasPrefix(instanceType, "t2.") {
-			return false
+func (p *KarpenterNodePoolProvisioner) kubectlExecute(ctx context.Context, args []string, stdin string) (string, error) {
+	token, err := p.tokenSource.Token()
+	if err != nil {
+		return "", err
+	}
+
+	args = append([]string{
+		"kubectl",
+		fmt.Sprintf("--server=%s", p.cluster.APIServerURL),
+		fmt.Sprintf("--token=%s", token.AccessToken),
+	}, args...)
+	if stdin != "" {
+		args = append(args, "-f", "-")
+	}
+
+	newApplyCommand := func() *exec.Cmd {
+		cmd := exec.Command(args[0], args[1:]...)
+		// prevent kubectl to find the in-cluster config
+		cmd.Env = []string{}
+		return cmd
+	}
+	var output bytes.Buffer
+	applyManifest := func() error {
+		cmd := newApplyCommand()
+		if stdin != "" {
+			cmd.Stdin = strings.NewReader(stdin)
 		}
+		cmd.Stdout = &output
+		_, err := p.execManager.Run(ctx, p.logger, cmd)
+		return err
 	}
-	return true
+	err = backoff.Retry(applyManifest, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxApplyRetries))
+	if err != nil {
+		return "", err
+	}
+	return output.String(), nil
+}
+
+func (p *KarpenterNodePoolProvisioner) provisionNodePool(ctx context.Context, nodePool *api.NodePool, values map[string]interface{}) error {
+	template, err := p.generateNodePoolTemplate(nodePool, values, provisionersFileName)
+	if err != nil {
+		return err
+	}
+	_, err = p.kubectlExecute(ctx, []string{"apply"}, template)
+	if err != nil {
+		return errors.Wrapf(err, "kubectl apply failed for node pool %s", nodePool.Name)
+	}
+	return nil
+}
+
+func (p *KarpenterNodePoolProvisioner) Reconcile(ctx context.Context, updater updatestrategy.UpdateStrategy) error {
+	karpenterPools := p.cluster.KarpenterPools()
+	output, err := p.kubectlExecute(ctx, []string{"get", "provisioner", "--no-headers", "-o", "custom-columns=':metadata.name'"}, "")
```

Comment:
IMO Let's use a proper client-go for interacting with the API! We only use `kubectl` binary for the `apply` case because it depend on client-side logic in `kubectl`. For getting/deleting provisioners we don't need it.